src/compiler/extensions.ts

```diff
@@ -0,0 +1,379 @@
+
+    export namespace ExtensionKind {
+        export const SemanticLint: "semantic-lint" = "semantic-lint";
+        export type SemanticLint = "semantic-lint";
+        export const SyntacticLint: "syntactic-lint" = "syntactic-lint";
+        export type SyntacticLint = "syntactic-lint";
+        export const LanguageService: "language-service" = "language-service";
+        export type LanguageService = "language-service";
+    }
+    export type ExtensionKind = ExtensionKind.SemanticLint | ExtensionKind.SyntacticLint | ExtensionKind.LanguageService;
+
+    export interface ExtensionCollectionMap {
+        "syntactic-lint"?: SyntacticLintExtension[];
+        "semantic-lint"?: SemanticLintExtension[];
+        "language-service"?: LanguageServiceExtension[];
+        [index: string]: Extension[] | undefined;
+    }
+
+    export interface ExtensionBase {
+        name: string;
+        args: any;
+        kind: ExtensionKind;
+    }
+
+    export type Timestamp = number & { __timestampBrand: void };
+    export interface ProfileData {
+        globalBucket: string;
+        task: string;
+        start: Timestamp;
+        length?: Timestamp;
+    }
+
+    export interface SyntacticLintExtension extends ExtensionBase {
+        kind: ExtensionKind.SyntacticLint;
+        ctor: SyntacticLintProviderStatic;
+    }
+
+    export interface SemanticLintExtension extends ExtensionBase {
+        kind: ExtensionKind.SemanticLint;
+        ctor: SemanticLintProviderStatic;
+    }
+
+    export interface LanguageServiceExtension extends ExtensionBase {
+        kind: ExtensionKind.LanguageService;
+        ctor: LanguageServiceProviderStatic;
+    }
+
+    export type Extension = SyntacticLintExtension | SemanticLintExtension | LanguageServiceExtension;
+
+    export interface ExtensionCache {
+        getCompilerExtensions(): ExtensionCollectionMap;
+        getExtensionLoadingDiagnostics(): Diagnostic[];
+    }
+
+    export interface ExtensionHost extends ModuleResolutionHost {
+        loadExtension?(name: string): any;
+    }
+
+    export interface Program {
+        /**
+         * Gets a map of loaded compiler extensions
+         */
+        getCompilerExtensions(): ExtensionCollectionMap;
+    }
+
+    /* @internal */
+    export interface TypeCheckerHost {
+        getCompilerExtensions(): ExtensionCollectionMap;
+    }
+
+    function profileTrace(trace: (s: string) => void | undefined, message: DiagnosticMessage, ...args: any[]) {
+        if (trace) {
+            trace(flattenDiagnosticMessageText(createCompilerDiagnostic(message, ...args).messageText, (sys && sys.newLine || "\n")));
+        }
+    }
+
+    declare var performance: { now?(): number }; // If we're running in a context with high resolution timers, make use of them
+    declare var process: { hrtime?(start?: [number, number]): [number, number] };
+
+    function getTimestampInternal(): number {
+        if (typeof performance !== "undefined" && performance.now) {

```

Comment:
I don't know much Javascript style -- why can't you say `performance && performance.now` ?